backend/src/org/commcare/session/CommCareSession.java

```diff
@@ -317,7 +320,14 @@ private SessionDatum getFirstMissingDatum(OrderedHashtable datumsCollectedSoFar,
                                               Vector<SessionDatum> allDatumsNeeded) {
         for (SessionDatum datum : allDatumsNeeded) {
             if (!datumsCollectedSoFar.containsKey(datum.getDataId())) {
-                return datum;
+                if (nextDatumValue != null) {
+                    // set the needed datum to the return value from the last
+                    // popped frame and continue looking for needed datums
+                    setDatum(datum.getDataId(), nextDatumValue);
+                    nextDatumValue = null;
+                } else {
+                    return datum;
+                }
```

Comment:
Sorta feel like this is a bad place to be doing this work, but couldn't think of a better place